zebra/zebra_rnh.c

```diff
@@ -623,10 +628,34 @@ zebra_rnh_resolve_nexthop_entry(struct zebra_vrf *zvrf, afi_t afi,
 				break;
 		}
 
-		/* Route entry found, we're done; else, walk up the tree. */
+		/* Route entry found, we're done; else, walk up the tree.
+		 *
+		 * BGP routes are usually recursive, though. Keep resolving
+		 * on the nexthop until hitting a non-BGP route.
+		 */
 		if (re) {
-			*prn = rn;
-			return re;
+			if (re->type != ZEBRA_ROUTE_BGP) {
+				*prn = rn;
+				return re;
+			}
+
+			/*
+			 * Note an aggregate has the NEXTHOP_TYPE_BLACKHOLE
+			 * as the nexthop type.
+			 */
+			nexthop = re->nhe->nhg.nexthop;
+			if (!nexthop || ((nexthop->type != NEXTHOP_TYPE_IPV4) &&
+					 (nexthop->type != NEXTHOP_TYPE_IPV6))) {
+				*prn = rn;
+				return re;
+			}
+
+			addr2hostprefix(afi2family(afi), &nexthop->gate, &p);
+			if (IS_ZEBRA_DEBUG_NHT_DETAILED) {
+				zlog_debug("  Continue resolving on nexthop %pFX",
+					   &p);
+			}
+			goto resolve_again;
```

Comment:
remove the goto and do the lookup here again.  We already have a `while (rn)` loop let's use it.

Reply:
Will do.